### PR TITLE
Refactor quick bar to use a common interface for future commands and …

### PR DESCRIPTION
**Pull request by donkawechico**
## Proposed change

This PR refactors quick bar to use a common `QuickBarItem` interface rather than juggling two types of entity lists.

This sets us up nicely for three things:

1. New command items. No longer needs to be an HA service call. If you can put it into a callback, it can be added as a command.
2. Easier sorting once matcher sorts by score: https://github.com/home-assistant/frontend/pull/7367. Currently, any component using fuzzy filter has to duplicate a complex `sort()` where a numeric return sorts descending, string return sorts ascending, undefined return isn't included. With a common interface, we can make a single sorter.
3. Any changes that need to be made to the quickbar now only have to happen in one place rather than duplicated across two renderers.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [X] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by donkawechico** [comment]
Is it just me or is there now too much top padding?

File: src/dialogs/quick-bar/ha-quick-bar.ts [hunk 1]
```diff
@@ -226,69 +203,20 @@ export class QuickBar extends LitElement {
...
-
-  private _launchMoreInfoDialog(entityId) {
-    fireEvent(this, "hass-more-info", { entityId });
-    this.closeDialog();
+  private _generateEntityItems(): QuickBarItem[] {
+    return Object.keys(this.hass.states).map((entityId) => ({
+      text: computeStateName(this.hass.states[entityId]),
+      altText: entityId,
+      icon: domainIcon(computeDomain(entityId)),
+      action: async () => {
```
Comment (bramkragten) [inline src/dialogs/quick-bar/ha-quick-bar.ts:216]:
```suggestion
      action: () => {
```

File: src/dialogs/quick-bar/ha-quick-bar.ts [hunk 2]
```diff
@@ -226,69 +203,20 @@ export class QuickBar extends LitElement {
...
-  }
-
-  private _launchMoreInfoDialog(entityId) {
-    fireEvent(this, "hass-more-info", { entityId });
-    this.closeDialog();
+  private _generateEntityItems(): QuickBarItem[] {
+    return Object.keys(this.hass.states).map((entityId) => ({
+      text: computeStateName(this.hass.states[entityId]),
+      altText: entityId,
+      icon: domainIcon(computeDomain(entityId)),
```
Comment (bramkragten) [inline src/dialogs/quick-bar/ha-quick-bar.ts:215]: Let's pass the state to `domainIcon`
```suggestion
      icon: domainIcon(computeDomain(entityId, this.hass.states[entityId])),
```

**Review by bramkragten — APPROVED** [review]

File: src/dialogs/quick-bar/ha-quick-bar.ts (hunk 1 above)
Comment (donkawechico) [reply src/dialogs/quick-bar/ha-quick-bar.ts:216]: This won't work. `QuickBarItem['action']` is a callback that expects a Promise return. This is so we can wait for actions to finish before closing the dialog

File: src/dialogs/quick-bar/ha-quick-bar.ts (hunk 2 above)
Comment (donkawechico) [reply src/dialogs/quick-bar/ha-quick-bar.ts:215]: Sure thing. What's that give us?

Looking at the code, it looks like it gives us better specificity on icons based on things like a binary sensor's classification?

Comment (bramkragten) [reply src/dialogs/quick-bar/ha-quick-bar.ts]: yep :-)

File: src/dialogs/quick-bar/ha-quick-bar.ts (hunk 1 above)
Comment (donkawechico) [reply src/dialogs/quick-bar/ha-quick-bar.ts]: We discussed on discord and @bramkragten coached me through the change (I can remove the Promise typing for `action` and still get the await behavior).

Good suggestion too as it made me discover that this branch had a regression in it removing progress spinner for awaited actions!

**Pull request merged**